proposals/0000-decl-invis-binders.rst

```diff
@@ -0,0 +1,621 @@
+
+1. Relax the syntactic check of ``data``, ``newtype``, ``type``, ``class``,
+   ``type family``, and ``data family`` declarations to allow ``@k``-binders in
+   their headers::
+
+     tv_bndr ::=
+              | tyvar                         -- variable
+              | '(' tyvar '::' kind ')'       -- variable with kind annotation
+       (NEW)  | '@' tyvar                     -- invisible variable
+       (NEW)  | '@' '(' tyvar '::' kind ')'   -- invisible variable with kind annotation
+       (NEW)  | '@' '_'                       -- wildcard (to skip an invisible quantifier)
+
+   The occurrences of ``@`` must be *prefix*, as defined by
+   `#229 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0229-whitespace-bang-patterns.rst>`_.
+
+2. When a declaration has no standalone kind signature, a ``@k``-binder gives
+   rise to a ``forall k.`` quantifier in the inferred kind signature.
+   The inferred ``forall k.`` does not float to the left; the order of
+   quantifiers continues to match the order of binders in the header.
+
+3. When a standalone kind signature is given, modify the process of zipping
+   binders with quantifiers to match ``@k``-binders and ``forall k.``-quantifiers.
+   To be precise, modify the pseudo-code of ``zipSAKS`` as follows::
+
+    -- Updated version
+    zipSAKS (q:qs) (b:bs)
+      | zippable q b  = (q, Just b)  : zipSAKS qs bs
+      | skippable q   = (q, Nothing) : zipSAKS qs (b:bs)
+      | otherwise     = error "Unzippable quantifier/binder pair"
+    zipSAKS [] (b:bs) = error "Too many binders"
+    zipSAKS _ [] = []
+
+    skippable q = isInvisibleQuantifier q
+    zippable q b =
+      (isInvisibleForall q && isInvisibleBinder b) ||
+      (isVisibleQuantifier q && isVisibleBinder b)
+
+   ``isInvisibleForall`` holds for the ``forall x.`` quantifier only, but does
+   not hold for ``forall {x}.`` or any other quantifier.
+
+   ``isInvisibleBinder`` holds for ``@k``, ``@(k :: s)``, and ``@_``, but does
+   not hold for ``k`` or ``(k :: s)``. ``isVisibleBinder`` is the opposite.
+
+   ``isVisibleQuantifier`` holds for ``a ->`` and ``forall x ->`` only.
+   ``isInvisibleQuantifier`` is the opposite.
+
+   This implies that within a single declaration, some ``forall x.``
+   quantifiers may be zipped with binders, while others are not.
+
+Secondary Change: Arity Inference
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+4. During arity inference, do not include trailing ``forall x.`` quantifiers in
+   the arity.
+
+Secondary Change: Scoping Rules
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+5. In type synonym declarations, require that every variable mentioned on the
+   RHS must be bound on the LHS.
+
+6. In type family and data family instances, require that every variable
+   mentioned on the RHS must also occur on the LHS.
+
+Secondary Change: Instantiation Rules
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+7. In type family and data family instances, the instantiation is fully
```

Comment:
This, similarly, is completely separable form the rest of the proposal -- even the other point about instances. They are in sympathy with each other (and, indeed, the rest of the proposal), but there is no dependency between the rest of this proposal and these two points about instances. Right?